File: app/src/main/java/life/plank/juna/zone/util/customview/PublicBoardToolbar.java
package life.plank.juna.zone.util.customview;

import android.app.Activity;
import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Point;
import android.os.CountDownTimer;
import android.os.SystemClock;
import android.util.AttributeSet;
import android.util.TypedValue;
import android.view.View;
import android.widget.Chronometer;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.request.RequestOptions;
import com.google.android.material.tabs.TabLayout;

import java.util.Date;
import java.util.TimeZone;

import androidx.annotation.DrawableRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.Toolbar;
import androidx.fragment.app.Fragment;
import androidx.viewpager.widget.ViewPager;
import life.plank.juna.zone.R;
import life.plank.juna.zone.ZoneApplication;
import life.plank.juna.zone.data.model.BoardTemperature;
import life.plank.juna.zone.data.model.MatchDetails;
import life.plank.juna.zone.interfaces.BoardHeaderListener;
import life.plank.juna.zone.interfaces.CustomViewListener;
import life.plank.juna.zone.interfaces.EngagementInfoTilesToolbar;
import life.plank.juna.zone.util.common.NumberFormatter;
import life.plank.juna.zone.util.facilis.ViewUtilKt;
import life.plank.juna.zone.util.time.DateUtil;

import static life.plank.juna.zone.util.common.AppConstants.BOARD_POPUP;
import static life.plank.juna.zone.util.common.AppConstants.FOUR_HOURS_MILLIS;
import static life.plank.juna.zone.util.common.AppConstants.FULL_TIME_LOWERCASE;
import static life.plank.juna.zone.util.common.AppConstants.GMT;
import static life.plank.juna.zone.util.common.AppConstants.LIVE;
import static life.plank.juna.zone.util.common.AppConstants.MatchTimeVal.MATCH_ABOUT_TO_START;
import static life.plank.juna.zone.util.common.AppConstants.MatchTimeVal.MATCH_ABOUT_TO_START_BOARD_ACTIVE;
import static life.plank.juna.zone.util.common.AppConstants.MatchTimeVal.MATCH_COMPLETED_TODAY;
import static life.plank.juna.zone.util.common.AppConstants.MatchTimeVal.MATCH_LIVE;
import static life.plank.juna.zone.util.common.AppConstants.MatchTimeVal.MATCH_PAST;
import static life.plank.juna.zone.util.common.AppConstants.MatchTimeVal.MATCH_SCHEDULED_LATER;
import static life.plank.juna.zone.util.common.AppConstants.MatchTimeVal.MATCH_SCHEDULED_TODAY;
import static life.plank.juna.zone.util.common.AppConstants.NS;
import static life.plank.juna.zone.util.common.AppConstants.ONE_HOUR_MILLIS;
import static life.plank.juna.zone.util.common.DataUtil.getDisplayTimeStatus;
import static life.plank.juna.zone.util.common.DataUtil.getSeparator;
import static life.plank.juna.zone.util.customview.CustomPopup.showOptionPopup;
import static life.plank.juna.zone.util.time.DateUtil.getHourMinuteSecondFormatDate;
import static life.plank.juna.zone.util.time.DateUtil.getMatchTimeValue;
import static life.plank.juna.zone.util.time.DateUtil.getScheduledMatchDateString;
import static life.plank.juna.zone.util.time.DateUtil.getTimeInFootballFormat;
import static life.plank.juna.zone.util.view.UIDisplayUtil.getDp;

public class PublicBoardToolbar extends Toolbar implements CustomViewListener, EngagementInfoTilesToolbar {

    private ImageView leagueLogoView;
    private LinearLayout scoreLayout;
    private TextView scoreView;
    private Chronometer timeStatusView;
    private ImageView winPointer;
    private ImageView homeTeamLogoView;
    private ImageView visitingTeamLogoView;
    private ImageButton optionsMenu;
    private ImageButton shareBtn;
    private TextView peopleCountView;
    private TextView postCountView;
    private TextView interactionCountView;
    private TabLayout infoTilesTabLayout;
    
    private boolean isFavourite;
    private boolean isNotificationOn;
    private boolean isFollowing;
    private int layoutRefreshState = -1;

    private int matchTimeValue;
    private long lastStopTime;
    private long baseTime;
    private MatchDetails matchDetails;
    private CountDownTimer countDownTimer;
    private BoardHeaderListener listener;

    public PublicBoardToolbar(Context context) {
        this(context, null);
    }

    public PublicBoardToolbar(Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public PublicBoardToolbar(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init(context, attrs);
    }

    private void init(Context context, AttributeSet attrs) {
        initViews(inflate(context, R.layout.public_board_toolbar, this));
        TypedArray array = context.obtainStyledAttributes(attrs, R.styleable.PublicBoardToolbar);

        boolean initWithDefaults = array.getBoolean(R.styleable.PublicBoardToolbar_useDefaults, true);
        if (initWithDefaults) {
            initWithDefaults(context);
            showLock(array.getBoolean(R.styleable.PublicBoardToolbar_showLock, false));
            return;
        }
        setScore(array.getString(R.styleable.PublicBoardToolbar_score));
        setLeagueLogo(array.getResourceId(R.styleable.PublicBoardToolbar_leagueLogo, 0));
        setHomeTeamLogo(array.getResourceId(R.styleable.PublicBoardToolbar_leagueLogo, 0));
        setVisitingTeamLogo(array.getResourceId(R.styleable.PublicBoardToolbar_leagueLogo, 0));
        setPeopleCount(array.getString(R.styleable.PublicBoardToolbar_peopleCount));
        setPostCount(array.getString(R.styleable.PublicBoardToolbar_postCount));
        setBoardTitle(array.getString(R.styleable.PublicBoardToolbar_boardTitle));
        showLock(array.getBoolean(R.styleable.PublicBoardToolbar_showLock, false));

        array.recycle();
    }

    //    TODO: Remove when converting this class to kotlin
    private void initViews(View rootView) {
        leagueLogoView = rootView.findViewById(R.id.logo);
        scoreLayout = rootView.findViewById(R.id.score_layout);
        scoreView = rootView.findViewById(R.id.score);
        timeStatusView = rootView.findViewById(R.id.time_status);
        winPointer = rootView.findViewById(R.id.win_pointer);
        homeTeamLogoView = rootView.findViewById(R.id.home_team_logo);
        visitingTeamLogoView = rootView.findViewById(R.id.visiting_team_logo);
        optionsMenu = rootView.findViewById(R.id.options_menu);
        shareBtn = rootView.findViewById(R.id.share_btn);
        peopleCountView = rootView.findViewById(R.id.people_count);
        postCountView = rootView.findViewById(R.id.post_count);
        interactionCountView = rootView.findViewById(R.id.interaction_count);
        infoTilesTabLayout = rootView.findViewById(R.id.info_tiles_tab_layout);
    }

    public void setUpPopUp(Activity activity, Long currentMatchId) {
        optionsMenu.setOnClickListener(view -> {
            int[] location = new int[2];

            view.getLocationOnScreen(location);

            //Initialize the Point with x, and y positions
            Point point = new Point();
            point.x = location[0];
            point.y = location[1];
            showOptionPopup(activity, point, BOARD_POPUP, currentMatchId, -400, 100);
        });
    }

    @Override
    public void initListeners(Fragment fragment) {
        if (fragment instanceof BoardHeaderListener) {
            listener = (BoardHeaderListener) fragment;
        } else throw new IllegalStateException("Fragment must implement BoardHeaderListener");

        addInfoTilesListener();
    }

    @Override
    public void initListeners(Activity activity) {
        if (activity instanceof BoardHeaderListener) {
            listener = (BoardHeaderListener) activity;
        } else throw new IllegalStateException("Activity must implement BoardHeaderListener");

        addInfoTilesListener();
    }

    @Override
    public void dispose() {
        listener = null;
        optionsMenu.setOnClickListener(null);
        switch (matchTimeValue) {
            case MATCH_LIVE:
                timeStatusView.stop();
                lastStopTime = SystemClock.elapsedRealtime() - (new Date().getTime() - matchDetails.getMatchStartTime().getTime());
                break;
            case MATCH_ABOUT_TO_START:
            case MATCH_SCHEDULED_LATER:
                resetCountDownTimer();
                break;
        }
    }

    private void addInfoTilesListener() {
        ViewUtilKt.onDebouncingClick(shareBtn, () -> {
            listener.onShareClick();
            return null;
        });
        switch (matchTimeValue) {
            case MATCH_LIVE:
                if (baseTime > 0) {
                    if (lastStopTime == 0) {
                        timeStatusView.setBase(baseTime);
                    } else {
                        long intervalOnPause = (baseTime - lastStopTime);
                        timeStatusView.setBase(timeStatusView.getBase() + intervalOnPause);
                    }
                    timeStatusView.start();
                }
                break;
            case MATCH_ABOUT_TO_START:
                setScheduledTimeStatus(true);
                break;
            case MATCH_SCHEDULED_LATER:
                setScheduledTimeStatus(false);
                break;
        }
    }

    private void initWithDefaults(Context context) {
        leagueLogoView.setImageResource(R.drawable.img_epl_logo);
        homeTeamLogoView.setImageResource(R.drawable.ic_arsenal_logo);
        visitingTeamLogoView.setImageResource(R.drawable.ic_blackpool_logo);
        scoreView.setText(context.getString(R.string._8_8));
    }

    public int getSelectedSection() {
        return infoTilesTabLayout.getSelectedTabPosition();
    }

    public void setScore(String score) {
        if (scoreLayout.getPaddingTop() > 0) {
            scoreLayout.setPadding(0, 0, 0, 0);
        }
        scoreView.setTextSize(TypedValue.COMPLEX_UNIT_SP, 32);
        scoreView.setText(score);
    }

    public void prepare(MatchDetails matchDetails, @DrawableRes int leagueLogo) {
        this.matchDetails = matchDetails;
        setLeagueLogo(leagueLogo);
        setHomeTeamLogo(matchDetails.getHomeTeam().getLogoLink());
        setVisitingTeamLogo(matchDetails.getAwayTeam().getLogoLink());
        setScore(getSeparator(matchDetails, winPointer, true));
        setBoardTitle(ZoneApplication.getContext().getString(R.string.matchday_) + matchDetails.getMatchDay());
        switch (getMatchTimeValue(matchDetails.getMatchStartTime(), false)) {
            case MATCH_PAST:
            case MATCH_COMPLETED_TODAY:
                setFullTimeStatus();
                matchTimeValue = MATCH_PAST;
                break;
            case MATCH_LIVE:
                setLiveTimeStatus(matchDetails.getMatchStartTime(), matchDetails.getTimeStatus());
                matchTimeValue = MATCH_LIVE;
                break;
            case MATCH_ABOUT_TO_START:
            case MATCH_ABOUT_TO_START_BOARD_ACTIVE:
            case MATCH_SCHEDULED_TODAY:
                setScheduledTimeStatus(true);
                setTodayMatchCountdown();
                matchTimeValue = MATCH_ABOUT_TO_START;
                break;
            case MATCH_SCHEDULED_LATER:
                setScheduledTimeStatus(false);
                matchTimeValue = MATCH_SCHEDULED_LATER;
                break;
        }
    }

    /**
     * Time status setter for past matches
     */
    public void setFullTimeStatus() {
        this.timeStatusView.setText(FULL_TIME_LOWERCASE);
    }

    /**
     * Time status setter for live matches
     */
    public void setLiveTimeStatus(Date matchStartTime, String timeStatus) {
        switch (timeStatus) {
            case NS:
            case LIVE:
                resetCountDownTimer();
                baseTime = SystemClock.elapsedRealtime() - (new Date().getTime() - matchStartTime.getTime());
                Date baseDate = new Date(baseTime);
                timeStatusView.setOnChronometerTickListener(chronometer -> {
                    baseTime += 1000;
                    baseDate.setTime(baseTime);
                    timeStatusView.setText(getTimeInFootballFormat(baseDate));
                });
                timeStatusView.start();
                break;
            default:
                timeStatusView.stop();
                timeStatusView.setText(getDisplayTimeStatus(timeStatus));
                break;
        }
    }

    /**
     * Time status setter for live matches
     */
    public void setTodayMatchCountdown() {
        long timeDiffFromNow = Math.abs(matchDetails.getMatchStartTime().getTime() - new Date().getTime());
        resetCountDownTimer();
        DateUtil.HOUR_MINUTE_SECOND_DATE_FORMAT.setTimeZone(TimeZone.getTimeZone(GMT));
        countDownTimer = new CountDownTimer(timeDiffFromNow, 1000) {
            @Override
            public void onTick(long millisUntilFinished) {
                timeStatusView.setText(getHourMinuteSecondFormatDate(new Date(millisUntilFinished)));
                if (millisUntilFinished <= FOUR_HOURS_MILLIS && millisUntilFinished > ONE_HOUR_MILLIS && layoutRefreshState == -1) {
                    layoutRefreshState = MATCH_ABOUT_TO_START_BOARD_ACTIVE;
                    listener.onMatchTimeStateChange();
                } else if (millisUntilFinished <= ONE_HOUR_MILLIS && layoutRefreshState == MATCH_ABOUT_TO_START_BOARD_ACTIVE) {
                    layoutRefreshState = MATCH_ABOUT_TO_START;
                    listener.onMatchTimeStateChange();
                }
            }

            @Override
            public void onFinish() {
                timeStatusView.setText(LIVE);
                matchDetails.setTimeStatus(LIVE);
                setLiveTimeStatus(matchDetails.getMatchStartTime(), matchDetails.getTimeStatus());
                setScore(getSeparator(matchDetails, winPointer, true));
                listener.onMatchTimeStateChange();
            }
        };
        countDownTimer.start();
    }

    /**
     * Time status setter for scheduled matches
     */
    public void setScheduledTimeStatus(boolean isScheduledToday) {
        winPointer.setVisibility(GONE);
        scoreView.setVisibility(isScheduledToday ? VISIBLE : GONE);
        scoreView.setTextSize(TypedValue.COMPLEX_UNIT_SP, isScheduledToday ? 10 : 32);
        scoreLayout.setPadding(0, (int) getDp(9), 0, 0);
        String scheduledMatchDateString = getScheduledMatchDateString(matchDetails.getMatchStartTime());
        if (isScheduledToday) {
//            set time of match in score view and set countdown in time status view
            scoreView.setText(scheduledMatchDateString);
        } else {
//            set time of match in time status view
            timeStatusView.setText(scheduledMatchDateString);
        }
    }

    private void resetCountDownTimer() {
        if (countDownTimer != null) {
            countDownTimer.cancel();
        }
    }

    @Override
    public void setLeagueLogo(@NonNull String logoUrl) {
        Glide.with(this).load(logoUrl)
                .apply(RequestOptions.centerInsideTransform()
                        .override((int) getDp(30), (int) getDp(30))
                        .placeholder(R.drawable.ic_place_holder)
                        .error(R.drawable.ic_place_holder))
                .into(leagueLogoView);
    }

    @Override
    public void setLeagueLogo(@DrawableRes int resource) {
        if (resource != 0) leagueLogoView.setImageResource(resource);
    }

    public void setHomeTeamLogo(String logoUrl) {
        Glide.with(this).load(logoUrl)
                .apply(RequestOptions.centerInsideTransform()
                        .override((int) getDp(30), (int) getDp(30))
                        .placeholder(R.drawable.ic_place_holder)
                        .error(R.drawable.ic_place_holder))
                .into(homeTeamLogoView);
    }

    public void setHomeTeamLogo(@DrawableRes int resource) {
        if (resource != 0) homeTeamLogoView.setImageResource(resource);
    }

    public void setVisitingTeamLogo(String logoUrl) {
        Glide.with(this).load(logoUrl)
                .apply(RequestOptions.centerInsideTransform()
                        .override((int) getDp(30), (int) getDp(30))
                        .placeholder(R.drawable.ic_place_holder)
                        .error(R.drawable.ic_place_holder))
                .into(visitingTeamLogoView);
    }

    public void setVisitingTeamLogo(@DrawableRes int resource) {
        if (resource != 0) visitingTeamLogoView.setImageResource(resource);
    }

    @Override
    public void setPeopleCount(@NonNull String peopleCount) {
        peopleCountView.setText(peopleCount);
    }

    @Override
    public void setPostCount(@NonNull String postCount) {
        postCountView.setText(postCount);
    }

    @Override
    public void setBoardTitle(@NonNull String boardTitle) { }

    @Override
    public void showLock(boolean showLock) { }

    @NonNull
    @Override
    public TabLayout getInfoTilesTabLayout() {
        return infoTilesTabLayout;
    }

    @Override
    public void setupWithViewPager(@NonNull ViewPager viewPager, int defaultSelection) {
        infoTilesTabLayout.setupWithViewPager(viewPager);
        viewPager.setCurrentItem(defaultSelection, true);
    }

    public void setBoardTemperature(BoardTemperature boardTemperature) {
        peopleCountView.setText(NumberFormatter.format(boardTemperature.getUserCount()));
        postCountView.setText(NumberFormatter.format(boardTemperature.getPostCount()));
        interactionCountView.setText(NumberFormatter.format(boardTemperature.getInteractionCount()));
    }
}

File: app/src/main/java/life/plank/juna/zone/util/customview/PublicBoardToolbar.kt
package life.plank.juna.zone.util.customview

import android.app.Activity
import android.content.Context
import android.graphics.Point
import android.os.*
import android.util.*
import android.view.View
import android.widget.ImageView
import androidx.annotation.DrawableRes
import androidx.appcompat.widget.Toolbar
import androidx.fragment.app.Fragment
import androidx.viewpager.widget.ViewPager
import butterknife.ButterKnife
import com.bumptech.glide.Glide
import com.bumptech.glide.request.RequestOptions
import com.google.android.material.tabs.TabLayout
import com.prembros.facilis.util.onDebouncingClick
import kotlinx.android.synthetic.main.layout_board_engagement.view.*
import kotlinx.android.synthetic.main.public_board_toolbar.view.*
import life.plank.juna.zone.*
import life.plank.juna.zone.data.model.MatchDetails
import life.plank.juna.zone.interfaces.*
import life.plank.juna.zone.util.common.AppConstants.*
import life.plank.juna.zone.util.common.AppConstants.MatchTimeVal.*
import life.plank.juna.zone.util.common.JunaDataUtil.*
import life.plank.juna.zone.util.customview.CustomPopup.showOptionPopup
import life.plank.juna.zone.util.time.*
import life.plank.juna.zone.util.time.DateUtil.*
import life.plank.juna.zone.util.view.UIDisplayUtil.*
import java.util.*

class PublicBoardToolbar @JvmOverloads constructor(
        context: Context,
        attrs: AttributeSet? = null,
        defStyleAttr: Int = 0
) : Toolbar(context, attrs, defStyleAttr), CustomViewListener, EngagementInfoTilesToolbar {

    private var layoutRefreshState = -1

    private var matchTimeValue: Int = 0
    private var lastStopTime: Long = 0
    private var baseTime: Long = 0
    private lateinit var matchDetails: MatchDetails
    private var countDownTimer: CountDownTimer? = null
    private var listener: BoardHeaderListener? = null

    init {
        init(context, attrs)
    }

    private fun init(context: Context, attrs: AttributeSet?) {
        val rootView = View.inflate(context, R.layout.public_board_toolbar, this)
        ButterKnife.bind(this, rootView)
        val array = context.obtainStyledAttributes(attrs, R.styleable.PublicBoardToolbar)

        val initWithDefaults = array.getBoolean(R.styleable.PublicBoardToolbar_useDefaults, true)
        if (initWithDefaults) {
            initWithDefaults(context)
            showLock(array.getBoolean(R.styleable.PublicBoardToolbar_showLock, false))
            return
        }
        setScore(array.getString(R.styleable.PublicBoardToolbar_score)?.run { this } ?: "0 - 0")
        setLeagueLogo(array.getResourceId(R.styleable.PublicBoardToolbar_leagueLogo, 0))
        setTeamLogo(array.getResourceId(R.styleable.PublicBoardToolbar_leagueLogo, 0), home_team_logo)
        setTeamLogo(array.getResourceId(R.styleable.PublicBoardToolbar_leagueLogo, 0), visiting_team_logo)
        setPeopleCount(array.getString(R.styleable.PublicBoardToolbar_peopleCount)!!)
        setPostCount(array.getString(R.styleable.PublicBoardToolbar_postCount)!!)
        setBoardTitle(array.getString(R.styleable.PublicBoardToolbar_boardTitle)!!)
        showLock(array.getBoolean(R.styleable.PublicBoardToolbar_showLock, false))

        array.recycle()
    }

    fun prepare(matchDetails: MatchDetails, @DrawableRes leagueLogo: Int) {
        this.matchDetails = matchDetails
        setLeagueLogo(leagueLogo)
        setTeamLogo(matchDetails.homeTeam.logoLink, home_team_logo)
        setTeamLogo(matchDetails.awayTeam.logoLink, visiting_team_logo)
        setScore(getSeparator(matchDetails, win_pointer, true))
        setBoardTitle(ZoneApplication.getContext().getString(R.string.matchday_) + matchDetails.matchDay)
        when (getMatchTimeValue(matchDetails.matchStartTime, false)) {
            MATCH_PAST, MATCH_COMPLETED_TODAY -> {
                setFullTimeStatus()
                matchTimeValue = MATCH_PAST
            }
            MATCH_LIVE -> {
                setLiveTimeStatus(matchDetails.matchStartTime, matchDetails.timeStatus)
                matchTimeValue = MATCH_LIVE
            }
            MATCH_ABOUT_TO_START, MATCH_ABOUT_TO_START_BOARD_ACTIVE, MATCH_SCHEDULED_TODAY -> {
                setScheduledTimeStatus(true)
                setTodayMatchCountdown()
                matchTimeValue = MATCH_ABOUT_TO_START
            }
            MATCH_SCHEDULED_LATER -> {
                setScheduledTimeStatus(false)
                matchTimeValue = MATCH_SCHEDULED_LATER
            }
        }
    }

    fun setUpPopUp(activity: Activity, currentMatchId: Long?) {
        options_menu.setOnClickListener { view ->
            val location = IntArray(2)

            view.getLocationOnScreen(location)

            //Initialize the Point with x, and y positions
            val point = Point()
            point.x = location[0]
            point.y = location[1]
            showOptionPopup(activity, point, BOARD_POPUP, currentMatchId, -400, 100)
        }
    }

    override fun initListeners(fragment: Fragment) {
        if (fragment is BoardHeaderListener) {
            listener = fragment
        } else
            throw IllegalStateException("Fragment must implement PublicBoardHeaderListener")

        addInfoTilesListener()
    }

    override fun initListeners(activity: Activity) {
        if (activity is BoardHeaderListener) {
            listener = activity
        } else
            throw IllegalStateException("Activity must implement PublicBoardHeaderListener")

        addInfoTilesListener()
    }

    private fun addInfoTilesListener() {
        share_btn.onDebouncingClick { listener?.onShareClick() }

        when (matchTimeValue) {
            MATCH_LIVE -> {
            }
            MATCH_ABOUT_TO_START -> setScheduledTimeStatus(true)
            MATCH_SCHEDULED_LATER -> setScheduledTimeStatus(false)
        }
    }

    private fun initWithDefaults(context: Context) {
        league_logo.setImageResource(R.drawable.img_epl_logo)
        home_team_logo.setImageResource(R.drawable.ic_arsenal_logo)
        visiting_team_logo.setImageResource(R.drawable.ic_blackpool_logo)
        score_text_view.text = context.getString(R.string._8_8)
    }

    override fun dispose() {
        listener = null
        options_menu.setOnClickListener(null)
        when (matchTimeValue) {
            MATCH_LIVE -> {
                time_status.stop()
                lastStopTime = SystemClock.elapsedRealtime() - (Date().time - matchDetails.matchStartTime.time)
            }
            MATCH_ABOUT_TO_START, MATCH_SCHEDULED_LATER -> resetCountDownTimer()
        }
    }

    fun setScore(score: String) {
        if (score_layout.paddingTop > 0) {
            score_layout.setPadding(0, 0, 0, 0)
        }
        score_text_view.setTextSize(TypedValue.COMPLEX_UNIT_SP, 32f)
        score_text_view.text = score
    }

    /**
     * Time status setter for past matches
     */
    private fun setFullTimeStatus() {
        this.time_status.text = FULL_TIME_LOWERCASE
    }

    /**
     * Time status setter for live matches
     */
    fun setLiveTimeStatus(matchStartTime: Date, timeStatus: String?) {
        when (timeStatus) {
            NS, LIVE -> {
                resetCountDownTimer()
                baseTime = Date().time - matchStartTime.time
                time_status.setOnChronometerTickListener {
                    baseTime += 1000
                    time_status.text = getFootballTimeElapsed(baseTime)
                }
                time_status.start()
            }
            else -> {
                time_status.stop()
                time_status.text = getDisplayTimeStatus(timeStatus)
            }
        }
    }

    /**
     * Time status setter for live matches
     */
    private fun setTodayMatchCountdown() {
        val timeDiffFromNow = Math.abs(matchDetails.matchStartTime.time - Date().time)
        resetCountDownTimer()
        DateUtil.HOUR_MINUTE_SECOND_DATE_FORMAT.timeZone = TimeZone.getTimeZone(GMT)
        countDownTimer = object : CountDownTimer(timeDiffFromNow, 1000) {
            override fun onTick(millisUntilFinished: Long) {
                time_status.text = getHourMinuteSecondFormatDate(Date(millisUntilFinished))
                if (millisUntilFinished in (ONE_HOUR_MILLIS + 1)..FOUR_HOURS_MILLIS && layoutRefreshState == -1) {
                    layoutRefreshState = MATCH_ABOUT_TO_START_BOARD_ACTIVE
                    listener?.onMatchTimeStateChange()
                } else if (millisUntilFinished <= ONE_HOUR_MILLIS && layoutRefreshState == MATCH_ABOUT_TO_START_BOARD_ACTIVE) {
                    layoutRefreshState = MATCH_ABOUT_TO_START
                    listener?.onMatchTimeStateChange()
                }
            }

            override fun onFinish() {
                time_status.text = LIVE
                matchDetails.timeStatus = LIVE
                setLiveTimeStatus(matchDetails.matchStartTime, matchDetails.timeStatus)
                setScore(getSeparator(matchDetails, win_pointer, true))
                listener?.onMatchTimeStateChange()
            }
        }
        countDownTimer?.start()
    }

    /**
     * Time status setter for scheduled matches
     */
    private fun setScheduledTimeStatus(isScheduledToday: Boolean) {
        win_pointer.visibility = View.GONE
        score_text_view.visibility = if (isScheduledToday) View.VISIBLE else View.GONE
        score_text_view.setTextSize(TypedValue.COMPLEX_UNIT_SP, (if (isScheduledToday) 10 else 32).toFloat())
        score_layout.setPadding(0, getDp(9f).toInt(), 0, 0)
        val scheduledMatchDateString = getScheduledMatchDateString(matchDetails.matchStartTime)
        if (isScheduledToday) {
//            set time of match in score view and set countdown in time status view
            score_text_view.text = scheduledMatchDateString
        } else {
//            set time of match in time status view
            time_status.text = scheduledMatchDateString
        }
    }

    private fun resetCountDownTimer() {
        countDownTimer?.cancel()
    }

    override fun setLeagueLogo(logoUrl: String) {
        Glide.with(this).load(logoUrl)
                .apply(RequestOptions.centerInsideTransform()
                        .override(getDp(30f).toInt(), getDp(30f).toInt())
                        .placeholder(R.drawable.ic_place_holder)
                        .error(R.drawable.ic_place_holder))
                .into(league_logo)
    }

    override fun setLeagueLogo(@DrawableRes resource: Int) {
        if (resource != 0) league_logo.setImageDrawable(findDrawable(resource))
    }

    private fun setTeamLogo(logoUrl: String, logoImageView: ImageView) {
        Glide.with(this).load(logoUrl)
                .apply(RequestOptions.centerInsideTransform()
                        .override(getDp(30f).toInt(), getDp(30f).toInt())
                        .placeholder(R.drawable.ic_place_holder)
                        .error(R.drawable.ic_place_holder))
                .into(logoImageView)
    }

    private fun setTeamLogo(@DrawableRes resource: Int, logoImageView: ImageView) {
        if (resource != 0) logoImageView.setImageResource(resource)
    }

    override fun setPeopleCount(peopleCount: String) {
        people_count.text = peopleCount
    }

    override fun setPostCount(postCount: String) {
        post_count.text = postCount
    }

    override fun setBoardTitle(boardTitle: String) {}

    override fun showLock(showLock: Boolean) {}

    override fun getInfoTilesTabLayout(): TabLayout {
        return info_tiles_tab_layout
    }

    override fun setupWithViewPager(viewPager: ViewPager, defaultSelection: Int) {
        info_tiles_tab_layout.setupWithViewPager(viewPager)
        viewPager.setCurrentItem(defaultSelection, true)
    }

//    TODO: un-comment if required
/*    fun setBoardTemperature(boardTemperature: BoardTemperature) {
        people_count.text = NumberFormatter.format(boardTemperature.userCount)
        post_count.text = NumberFormatter.format(boardTemperature.postCount)
        interaction_count.text = NumberFormatter.format(boardTemperature.interactionCount)
    }*/
}